File: scripts/performance.js
#!/usr/bin/env node

import fs from 'node:fs/promises'
import { performance } from 'node:perf_hooks'
import process from 'node:process'
import Table from 'cli-table3'
import pc from 'picocolors'

const MODE = process.env.BENCHMARK_MODE || 'development'
const IS_PRODUCTION = MODE === 'production'
const RARI_PORT = IS_PRODUCTION ? 3000 : 5173
const NEXTJS_PORT = IS_PRODUCTION ? 3001 : 3000
const WARMUP_REQUESTS = 5
const TEST_REQUESTS = 10

const SCENARIOS = [
  { path: '/', name: 'Homepage (All Components)' },
]

async function measureRequest(url, requests = TEST_REQUESTS) {
  const times = []
  const sizes = []
  let errors = 0

  for (let i = 0; i < WARMUP_REQUESTS; i++) {
    try {
      await fetch(url)
    }
    catch {
    }
  }

  console.warn(`  Testing ${url}...`)

  for (let i = 0; i < requests; i++) {
    const start = performance.now()

    try {
      const response = await fetch(url)
      const end = performance.now()

      if (response.ok) {
        const text = await response.text()
        times.push(end - start)
        sizes.push(text.length)
      }
      else {
        errors++
      }
    }
    catch {
      errors++
    }

    await new Promise(resolve => setTimeout(resolve, 50))
  }

  if (times.length === 0) {
    return null
  }

  const sortedTimes = times.sort((a, b) => a - b)
  const avgSize = sizes.reduce((a, b) => a + b, 0) / sizes.length

  return {
    min: Math.min(...times),
    max: Math.max(...times),
    avg: times.reduce((a, b) => a + b, 0) / times.length,
    p50: sortedTimes[Math.floor(sortedTimes.length * 0.5)],
    p95: sortedTimes[Math.floor(sortedTimes.length * 0.95)],
    p99: sortedTimes[Math.floor(sortedTimes.length * 0.99)],
    avgSize: Math.round(avgSize),
    errors,
    successRate: ((requests - errors) / requests) * 100,
  }
}

async function benchmarkFramework(name, port, color) {
  console.warn(pc[color](`\n🚀 Benchmarking ${name} (port ${port})`))

  const results = {}

  for (const scenario of SCENARIOS) {
    const url = `http://localhost:${port}${scenario.path}`
    console.warn(`\n📊 ${scenario.name}`)

    const result = await measureRequest(url)
    if (result) {
      results[scenario.name] = result
      console.warn(`  ✅ Avg: ${result.avg.toFixed(2)}ms, P95: ${result.p95.toFixed(2)}ms, Size: ${result.avgSize}b`)
    }
    else {
      results[scenario.name] = { error: 'Failed to get valid responses' }
      console.warn(`  ❌ Failed to get valid responses`)
    }
  }

  return results
}

function displayComparison(rariResults, nextjsResults) {
  console.warn(pc.bold('\n📈 Performance Comparison'))

  const table = new Table({
    head: ['Scenario', 'Rari (ms)', 'Next.js (ms)', 'Difference', 'Winner'],
    colWidths: [20, 15, 15, 15, 10],
  })

  for (const scenario of SCENARIOS) {
    const rari = rariResults[scenario.name]
    const nextjs = nextjsResults[scenario.name]

    if (rari?.avg && nextjs?.avg) {
      const diff = ((rari.avg - nextjs.avg) / nextjs.avg) * 100
      const winner = rari.avg < nextjs.avg ? '🦀 Rari' : '🟢 Next.js'
      const diffStr = diff > 0 ? `+${diff.toFixed(1)}%` : `${diff.toFixed(1)}%`
      const diffColor = diff < 0 ? pc.green : pc.red

      table.push([
        scenario.name,
        rari.avg.toFixed(2),
        nextjs.avg.toFixed(2),
        diffColor(diffStr),
        winner,
      ])
    }
    else {
      table.push([
        scenario.name,
        rari?.error ? 'Error' : 'N/A',
        nextjs?.error ? 'Error' : 'N/A',
        'N/A',
        'N/A',
      ])
    }
  }

  console.warn(table.toString())
}

async function calculateSummary(rariResults, nextjsResults) {
  const validScenarios = SCENARIOS.filter(s =>
    rariResults[s.name]?.avg && nextjsResults[s.name]?.avg,
  )

  if (validScenarios.length === 0) {
    console.warn(pc.red('\n❌ No valid results to compare'))
    return
  }

  const rariAvg = validScenarios.reduce((sum, s) =>
    sum + rariResults[s.name].avg, 0) / validScenarios.length
  const nextjsAvg = validScenarios.reduce((sum, s) =>
    sum + nextjsResults[s.name].avg, 0) / validScenarios.length

  const improvement = ((nextjsAvg - rariAvg) / nextjsAvg) * 100

  console.warn(pc.bold('\n📊 Summary'))
  console.warn(`Average Response Time:`)
  console.warn(`  🦀 Rari:     ${rariAvg.toFixed(2)}ms`)
  console.warn(`  🟢 Next.js:  ${nextjsAvg.toFixed(2)}ms`)

  if (improvement > 0) {
    console.warn(pc.green(`  📈 Rari is ${improvement.toFixed(1)}% faster`))
  }
  else {
    console.warn(pc.red(`  📉 Rari is ${Math.abs(improvement).toFixed(1)}% slower`))
  }
}

async function saveResults(rariResults, nextjsResults) {
  const timestamp = new Date().toISOString()
  const results = {
    timestamp,
    rari: rariResults,
    nextjs: nextjsResults,
    summary: {
      testRequests: TEST_REQUESTS,
      warmupRequests: WARMUP_REQUESTS,
      scenarios: SCENARIOS.length,
    },
  }

  try {
    await fs.mkdir('./results', { recursive: true })
  }
  catch {
  }

  await fs.writeFile(
    `./results/performance-${timestamp.slice(0, 10)}.json`,
    JSON.stringify(results, null, 2),
  )

  await fs.writeFile(
    './results/latest.json',
    JSON.stringify(results, null, 2),
  )

  console.warn(pc.dim(`\n💾 Results saved to ./results/performance-${timestamp.slice(0, 10)}.json`))
}

async function main() {
  console.warn(pc.bold(pc.cyan('🏁 Rari vs Next.js Performance Benchmark')))
  console.warn(pc.dim('This benchmark compares server-side rendering performance\n'))

  console.warn(pc.blue(`📊 Running in ${pc.bold(MODE.toUpperCase())} mode`))
  console.warn('')

  console.warn(pc.yellow('⚠️  Make sure both applications are running:'))
  if (IS_PRODUCTION) {
    console.warn(`  🦀 Rari:     http://localhost:${RARI_PORT} (run: pnpm start)`)
    console.warn(`  🟢 Next.js:  http://localhost:${NEXTJS_PORT} (run: PORT=3001 pnpm start)`)
  }
  else {
    console.warn(`  🦀 Rari:     http://localhost:${RARI_PORT} (run: pnpm dev)`)
    console.warn(`  🟢 Next.js:  http://localhost:${NEXTJS_PORT} (run: pnpm dev)`)
  }
  console.warn('')

  try {
    await fetch(`http://localhost:${RARI_PORT}`)
    console.warn(pc.green('✅ Rari server is responding'))
  }
  catch {
    const command = IS_PRODUCTION ? 'pnpm start' : 'pnpm dev'
    console.warn(pc.red(`❌ Rari server is not responding. Please start it with: cd rari-app && ${command}`))
    process.exit(1)
  }

  try {
    await fetch(`http://localhost:${NEXTJS_PORT}`)
    console.warn(pc.green('✅ Next.js server is responding'))
  }
  catch {
    const command = IS_PRODUCTION ? 'pnpm start' : 'pnpm dev'
    console.warn(pc.red(`❌ Next.js server is not responding. Please start it with: cd nextjs-app && ${command}`))
    process.exit(1)
  }

  console.warn(pc.dim('\nStarting benchmark in 3 seconds...'))
  await new Promise(resolve => setTimeout(resolve, 3000))

  const rariResults = await benchmarkFramework('Rari', RARI_PORT, 'yellow')
  const nextjsResults = await benchmarkFramework('Next.js', NEXTJS_PORT, 'green')

  displayComparison(rariResults, nextjsResults)
  await calculateSummary(rariResults, nextjsResults)
  await saveResults(rariResults, nextjsResults)

  console.warn(pc.bold(pc.green('\n🎉 Benchmark completed!')))
}

main().catch(console.error)

File: tools/benchmark/src/performance.rs
mod time_utils;

use anyhow::{Context, Result};
use clap::Parser;
use colored::Colorize;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::time::{Duration, Instant, SystemTime};
use tabled::{Table, Tabled};
use tokio::fs;

#[derive(Parser, Debug)]
#[command(name = "performance")]
#[command(about = "Run performance benchmarks comparing rari and Next.js")]
struct Args {
    #[arg(short, long, default_value = "50")]
    warmup: usize,
    #[arg(short, long, default_value = "20")]
    requests: usize,
    #[arg(long, default_value = "3000")]
    rari_port: u16,
    #[arg(long, default_value = "3001")]
    nextjs_port: u16,
    #[arg(long, default_value = "results")]
    results_dir: PathBuf,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PerformanceMetrics {
    min: f64,
    max: f64,
    avg: f64,
    p50: f64,
    p95: f64,
    p99: f64,
    #[serde(rename = "avgSize")]
    avg_size: usize,
    errors: usize,
    #[serde(
        rename = "successRate",
        serialize_with = "time_utils::serialize_float_as_int_if_whole"
    )]
    success_rate: f64,
}

#[derive(Debug, Serialize, Deserialize)]
struct BenchmarkResults {
    timestamp: String,
    rari: HashMap<String, PerformanceMetrics>,
    nextjs: HashMap<String, PerformanceMetrics>,
    summary: TestSummary,
}

#[derive(Debug, Serialize, Deserialize)]
struct TestSummary {
    #[serde(rename = "testRequests")]
    test_requests: usize,
    #[serde(rename = "warmupRequests")]
    warmup_requests: usize,
    scenarios: usize,
}

#[derive(Debug, Clone)]
struct Scenario {
    path: String,
    name: String,
}

#[derive(Tabled)]
struct ComparisonRow {
    #[tabled(rename = "Scenario")]
    scenario: String,
    #[tabled(rename = "rari (ms)")]
    rari_ms: String,
    #[tabled(rename = "Next.js (ms)")]
    nextjs_ms: String,
    #[tabled(rename = "Difference")]
    difference: String,
    #[tabled(rename = "Winner")]
    winner: String,
}

async fn check_server(name: &str, port: u16) -> Result<()> {
    let url = format!("http://localhost:{}", port);
    reqwest::get(&url)
        .await
        .context(format!("{} server is not responding at {}", name, url))?;
    println!("{} {} server is responding", "✅".green(), name);
    Ok(())
}

async fn measure_request(url: &str, warmup: usize, requests: usize) -> Result<PerformanceMetrics> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(10))
        .build()?;

    println!("  Testing {}...", url);

    for _ in 0..warmup {
        let _ = client.get(url).send().await;
        tokio::time::sleep(Duration::from_millis(10)).await;
    }

    let mut times = Vec::new();
    let mut sizes = Vec::new();
    let mut errors = 0;

    for _ in 0..requests {
        let start = Instant::now();

        match client.get(url).send().await {
            Ok(response) if response.status().is_success() => {
                let elapsed = start.elapsed().as_secs_f64() * 1000.0;

                match response.text().await {
                    Ok(text) => {
                        times.push(elapsed);
                        sizes.push(text.len());
                    }
                    Err(_) => errors += 1,
                }
            }
            _ => errors += 1,
        }

        tokio::time::sleep(Duration::from_millis(10)).await;
    }

    if times.is_empty() {
        anyhow::bail!("No successful requests");
    }

    let mut sorted_times = times.clone();
    sorted_times.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let sum: f64 = times.iter().sum();
    let avg = sum / times.len() as f64;
    let avg_size = sizes.iter().sum::<usize>() / sizes.len();
    let success_rate = ((requests - errors) as f64 / requests as f64) * 100.0;

    Ok(PerformanceMetrics {
        min: sorted_times[0],
        max: sorted_times[sorted_times.len() - 1],
        avg,
        p50: percentile(&sorted_times, 0.50),
        p95: percentile(&sorted_times, 0.95),
        p99: percentile(&sorted_times, 0.99),
        avg_size,
        errors,
        success_rate,
    })
}

fn percentile(sorted_data: &[f64], p: f64) -> f64 {
    let index = (p * (sorted_data.len() - 1) as f64) as usize;
    sorted_data[index]
}

async fn benchmark_framework(
    name: &str,
    port: u16,
    scenarios: &[Scenario],
    warmup: usize,
    requests: usize,
) -> Result<HashMap<String, PerformanceMetrics>> {
    println!(
        "\n{} Benchmarking {} (port {})",
        "🚀".bold(),
        name.bold(),
        port
    );

    let mut results = HashMap::new();

    for scenario in scenarios {
        let url = format!("http://localhost:{}{}", port, scenario.path);
        println!("\n📊 {}", scenario.name);

        match measure_request(&url, warmup, requests).await {
            Ok(metrics) => {
                println!(
                    "  {} Avg: {:.2}ms, P95: {:.2}ms, Size: {}b",
                    "✅".green(),
                    metrics.avg,
                    metrics.p95,
                    metrics.avg_size
                );
                results.insert(scenario.name.clone(), metrics);
            }
            Err(e) => {
                println!("  {} Failed: {}", "❌".red(), e);
            }
        }
    }

    Ok(results)
}

fn display_comparison(
    scenarios: &[Scenario],
    rari_results: &HashMap<String, PerformanceMetrics>,
    nextjs_results: &HashMap<String, PerformanceMetrics>,
) {
    println!("\n{}", "📈 Performance Comparison".bold());

    let mut rows = Vec::new();

    for scenario in scenarios {
        if let (Some(rari), Some(nextjs)) = (
            rari_results.get(&scenario.name),
            nextjs_results.get(&scenario.name),
        ) {
            let diff = ((rari.avg - nextjs.avg) / nextjs.avg) * 100.0;
            let winner = if rari.avg < nextjs.avg {
                "🦀 rari"
            } else {
                "🟢 Next.js"
            };
            let diff_str = if diff > 0.0 {
                format!("+{:.1}%", diff)
            } else {
                format!("{:.1}%", diff)
            };

            rows.push(ComparisonRow {
                scenario: scenario.name.clone(),
                rari_ms: format!("{:.2}", rari.avg),
                nextjs_ms: format!("{:.2}", nextjs.avg),
                difference: diff_str,
                winner: winner.to_string(),
            });
        }
    }

    let table = Table::new(rows).to_string();
    println!("\n{}", table);
}

fn calculate_summary(
    scenarios: &[Scenario],
    rari_results: &HashMap<String, PerformanceMetrics>,
    nextjs_results: &HashMap<String, PerformanceMetrics>,
) {
    let valid_scenarios: Vec<_> = scenarios
        .iter()
        .filter(|s| rari_results.contains_key(&s.name) && nextjs_results.contains_key(&s.name))
        .collect();

    if valid_scenarios.is_empty() {
        println!("\n{}", "❌ No valid results to compare".red());
        return;
    }

    let rari_avg: f64 = valid_scenarios
        .iter()
        .map(|s| rari_results[&s.name].avg)
        .sum::<f64>()
        / valid_scenarios.len() as f64;

    let nextjs_avg: f64 = valid_scenarios
        .iter()
        .map(|s| nextjs_results[&s.name].avg)
        .sum::<f64>()
        / valid_scenarios.len() as f64;

    let improvement = ((nextjs_avg - rari_avg) / nextjs_avg) * 100.0;

    println!("\n{}", "📊 Summary".bold());
    println!("Average Response Time:");
    println!("  🦀 rari:     {:.2}ms", rari_avg);
    println!("  🟢 Next.js:  {:.2}ms", nextjs_avg);

    if improvement > 0.0 {
        println!("  {} rari is {:.1}% faster", "📈".green(), improvement);
    } else {
        println!("  {} rari is {:.1}% slower", "📉".red(), improvement.abs());
    }
}

async fn save_results(results: &BenchmarkResults, results_dir: &PathBuf) -> Result<()> {
    fs::create_dir_all(results_dir).await?;

    let now = SystemTime::now();
    let date = time_utils::format_date(now);
    let filename = results_dir.join(format!("performance-{}.json", date));

    let json = format!("{}\n", serde_json::to_string_pretty(results)?);
    fs::write(&filename, &json).await?;

    let latest = results_dir.join("latest.json");
    fs::write(&latest, &json).await?;

    println!(
        "\n{} Results saved to {}",
        "💾".dimmed(),
        filename.display()
    );

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    let args = Args::parse();

    println!(
        "{}",
        "🏁 rari vs Next.js Performance Benchmark".cyan().bold()
    );
    println!(
        "{}",
        "This benchmark compares server-side rendering performance\n".dimmed()
    );

    let scenarios = vec![Scenario {
        path: "/".to_string(),
        name: "Homepage (All Components)".to_string(),
    }];

    if let Err(e) = check_server("rari", args.rari_port).await {
        eprintln!("{} {}", "❌".red(), e);
        eprintln!("Please start the rari server with: cd rari-app && pnpm dev");
        std::process::exit(1);
    }

    if let Err(e) = check_server("Next.js", args.nextjs_port).await {
        eprintln!("{} {}", "❌".red(), e);
        eprintln!("Please start the Next.js server with: cd nextjs-app && pnpm dev");
        std::process::exit(1);
    }

    println!("\n{}", "Starting benchmark in 3 seconds...".dimmed());
    tokio::time::sleep(Duration::from_secs(3)).await;

    let rari_results = benchmark_framework(
        "rari",
        args.rari_port,
        &scenarios,
        args.warmup,
        args.requests,
    )
    .await?;
    let nextjs_results = benchmark_framework(
        "Next.js",
        args.nextjs_port,
        &scenarios,
        args.warmup,
        args.requests,
    )
    .await?;

    display_comparison(&scenarios, &rari_results, &nextjs_results);
    calculate_summary(&scenarios, &rari_results, &nextjs_results);

    let results = BenchmarkResults {
        timestamp: time_utils::format_timestamp(SystemTime::now()),
        summary: TestSummary {
            test_requests: args.requests,
            warmup_requests: args.warmup,
            scenarios: scenarios.len(),
        },
        rari: rari_results,
        nextjs: nextjs_results,
    };

    save_results(&results, &args.results_dir).await?;

    println!("\n{}", "🎉 Benchmark completed!".green().bold());

    Ok(())
}
